### ISSUE TITLE
[Feature Request] Preload Quests by gpx route

### ISSUE THREAD

**Issue by Mannshoch** [westnordost/StreetComplete#473]
Allow to load a gpx file to get the Quests along that route. So you do not need Internet on tour and you could plan the Quests you wish to answer.

**Comment by callis** [comment]
from the recent weekyOSM Newsletter:
mmd wrote about a nice prototype demonstrating the power of the around: filter in Overpass API to show how you can download OSM data along a GPX track.
https://www.openstreetmap.org/user/mmd/diary/42055

**Comment by westnordost** [comment]
Sounds promising, but it's not released yet.

Also, my previous statement stands. Is not an OSM limitation but a limitation of this app.

**Comment by l-mb** [comment]
Is there any chance to revisit this option? I know it implies potentially downloading quite a bit of data ahead of the hike, but having to take my device out of airplane mode (conserving battery) and manually pre-loading the next couple of areas that I'm likely passing through is a bit of a hassle.

It'd help if Street Complete at least had the ability to load and show a track (similar to how it shows the path I took?), then I could - if it can't be easily automated - do all the preloading manually.

(For most hikes, at that point Street Complete could even be my primary map app, having planned the hike with a different tool.)

**Comment by westnordost** [comment]
Hmm, loading a GPX track and downloading data along it sounds like a feature that would be used only by few power users, as it requires first to draw up a GPX track in some other program, then loading it in StreetComplete.

So, I think this would be something that would rather be accepted as a PR in [SCEE](https://github.com/Helium314/SCEE), an extended edition of StreetComplete geared towards power-users.

It is written in Kotlin, too. I think after parsing a GPX, it could be solved like this:

Downloads in StreetComplete must be done in bounding boxes that line up with a slippy map tile rect of zoom 16. So, we need to find which tiles to download.
```
var tiles
var last_tile

for each point in gpx_track:
  var tile = tile in which point is located
  add tile to tiles
  laste_tile = tile

// (I think there is an algorithm for that already somewhere in the source code)
reduce tiles to the smallest number of possible tile rects

download all these tile rects
```

What to do about if two GPX points are several tiles away from each other? Could either ignore that (so a GPX track with two points - Berlin Airport to Madrid Airport does not download a straight path through all of Europe). Otherwise, 
```
val vector = tile - last_tile
val steps = length of vector
for each step in steps:
  val in_between_tile = tile + vector * step
  add in_between_tile to tiles
```

**Comment by Helium314** [comment]
Such functionality is more or less available in SCEE.

But loading along the track is rather limited, simply using the track bbox.
Since I don't want to ignore the default download size limit (danger of OOM crashes), download often needs to be done manually anyway. A PR for improvement would definitely be welcome though.

**Comment by Mannshoch** [comment]
I don't think loading along the track need much data. Along a Track by car I only need the data 15m left and right the trackpoint. Along the track by foot may a bit more. 
If I plan a longer trip e.g. through Europe. Than I do not need loading the whole track emediately.
A download also could happen in several steps with some timegap I'm between.

**Comment by mnalis** [comment]
> Such functionality is more or less available in SCEE.
> But loading along the track is rather limited, simply using the track bbox.

Yes, in [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) fork of StreetComplete, it is available under `Settings` / `Display` / `Display GPX track from file`, so @Mannshoch might want to try that (SCEE otherwise mostly looks just like StreetComplete unless one enables extra features, and it installs alongside regular StreetComplete and does not affect it, so should be easy to try out. 

> Since I don't want to ignore the default download size limit (danger of OOM crashes), download often needs to be done manually anyway. A PR for improvement would definitely be welcome though.

Yeah, last time I tried it (hoping to preload inter-city bus route so I can map during the ride) it bailed out with `Download area too big, please download manually`. But it displays selected GPX tracks just fine, and allows for manual downloads along the track (which is still tedious - thus I've suggested https://github.com/Helium314/SCEE/issues/471).

**Comment by mnalis** [comment]
- (ref. https://github.com/streetcomplete/StreetComplete/discussions/5345 with some code at https://github.com/streetcomplete/StreetComplete/compare/master...greuters:StreetComplete:gpx-download)

**Comment by TheArenator** [comment]
I just did a 600 mile road trip. It took forever to manually download the map data ahead of time. This doesn't have to be a feature just for power users, it could be a mainline feature. I only see it as making the act of surveying more efficient. I have hope. On the upside, my 600 mile roadtrip was from Sacramento to San Diego and I got nearly every road tagged with it's surface during that trip. It kept me busy during the drive (as a passenger) and time flew by.

**Comment by westnordost** [comment]
Downloading data for a 600 mile trip, wow. Hey, mind telling me what Android displays in 

Android Settings → Apps → StreetComplete → Storage & Cache → User data

?

> On the upside, my 600 mile roadtrip was from Sacramento to San Diego and I got nearly every road tagged with it's surface during that trip. 

Nice! Congratulations!

Regarding the feature, there was a pull request that got really far IIRC, unfortunately the guy who started it lost interest.

**Comment by TheArenator** [comment]
Here's what it says:
App size 146 MB
User data 1.78 GB
Cache 1.39 MB
Total 1.93 GB

**Comment by Helium314** [comment]
> Regarding the feature, there was a pull request that got really far IIRC, unfortunately the guy who started it lost interest.

I took the code from this PR and put it (I think in some simplified form) into SCEE. Downloading can  take a lot of time, but probably it's _much_ better than doing all that manually.

**Comment by westnordost** [comment]
With a composed based API? Simplified sounds good!

**Comment by Helium314** [comment]
I didn't check, but it's been in SCEE for about 2 years with barely enough maintainance to keep it working. So most likely not implemented in a platform-independent way.
